File: go/console/output/csv_test.go
package output_test

import (
	"bytes"
	"encoding/csv"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"hop.top/kit/go/console/output"
)

type csvRow struct {
	ID    string `table:"ID"`
	Name  string `table:"Name"`
	Score int    `table:"Score"`
}

func renderCSV(t *testing.T, data any, optPairs []string, cols []string) (string, error) {
	t.Helper()
	f, ok := output.Default.Lookup(output.CSV)
	require.True(t, ok, "csv formatter must be registered on Default")
	opts, err := output.ParseOptions(optPairs, f.Options())
	require.NoError(t, err)
	var buf bytes.Buffer
	if err := f.Render(&buf, data, opts, cols); err != nil {
		return buf.String(), err
	}
	return buf.String(), nil
}

func TestCSV_DefaultDelimiterWithHeader(t *testing.T) {
	data := []csvRow{
		{ID: "1", Name: "alpha", Score: 10},
		{ID: "2", Name: "beta", Score: 20},
	}
	out, err := renderCSV(t, data, nil, nil)
	require.NoError(t, err)

	lines := strings.Split(strings.TrimRight(out, "\n"), "\n")
	require.Len(t, lines, 3, "header + 2 rows")
	assert.Equal(t, "ID,Name,Score", lines[0])
	assert.Equal(t, "1,alpha,10", lines[1])
	assert.Equal(t, "2,beta,20", lines[2])
}

func TestCSV_DelimiterOverride(t *testing.T) {
	data := []csvRow{{ID: "1", Name: "alpha", Score: 10}}
	out, err := renderCSV(t, data, []string{"delimiter=;"}, nil)
	require.NoError(t, err)

	lines := strings.Split(strings.TrimRight(out, "\n"), "\n")
	require.Len(t, lines, 2)
	assert.Equal(t, "ID;Name;Score", lines[0])
	assert.Equal(t, "1;alpha;10", lines[1])
}

func TestCSV_NoHeader(t *testing.T) {
	data := []csvRow{
		{ID: "1", Name: "alpha", Score: 10},
		{ID: "2", Name: "beta", Score: 20},
	}
	out, err := renderCSV(t, data, []string{"no-header=true"}, nil)
	require.NoError(t, err)

	lines := strings.Split(strings.TrimRight(out, "\n"), "\n")
	require.Len(t, lines, 2, "no header, only 2 data rows")
	assert.Equal(t, "1,alpha,10", lines[0])
	assert.Equal(t, "2,beta,20", lines[1])
}

func TestCSV_QuoteAll(t *testing.T) {
	data := []csvRow{{ID: "1", Name: "alpha", Score: 10}}
	out, err := renderCSV(t, data, []string{"quote-all=true"}, nil)
	require.NoError(t, err)

	lines := strings.Split(strings.TrimRight(out, "\n"), "\n")
	require.Len(t, lines, 2)
	assert.Equal(t, `"ID","Name","Score"`, lines[0])
	assert.Equal(t, `"1","alpha","10"`, lines[1])
}

func TestCSV_QuoteAll_EscapesInternalQuote(t *testing.T) {
	type quoted struct {
		Note string `table:"Note"`
	}
	data := []quoted{{Note: `say "hi"`}}
	out, err := renderCSV(t, data, []string{"quote-all=true"}, nil)
	require.NoError(t, err)

	lines := strings.Split(strings.TrimRight(out, "\n"), "\n")
	require.Len(t, lines, 2)
	assert.Equal(t, `"Note"`, lines[0])
	assert.Equal(t, `"say ""hi"""`, lines[1])
}

func TestCSV_CRLF(t *testing.T) {
	data := []csvRow{{ID: "1", Name: "alpha", Score: 10}}
	out, err := renderCSV(t, data, []string{"crlf=true"}, nil)
	require.NoError(t, err)
	assert.Contains(t, out, "\r\n")
	// Both lines (header + row) end in CRLF.
	assert.Equal(t, 2, strings.Count(out, "\r\n"))
}

func TestCSV_CRLFWithQuoteAll(t *testing.T) {
	data := []csvRow{{ID: "1", Name: "alpha", Score: 10}}
	out, err := renderCSV(t, data, []string{"crlf=true", "quote-all=true"}, nil)
	require.NoError(t, err)
	assert.Equal(t, 2, strings.Count(out, "\r\n"))
	assert.Contains(t, out, `"ID","Name","Score"`+"\r\n")
}

func TestCSV_EmptySlice(t *testing.T) {
	out, err := renderCSV(t, []csvRow{}, nil, nil)
	require.NoError(t, err)
	assert.Empty(t, out, "empty slice produces no output (parity with table formatter)")
}

func TestCSV_ColsSubset(t *testing.T) {
	data := []csvRow{
		{ID: "1", Name: "alpha", Score: 10},
		{ID: "2", Name: "beta", Score: 20},
	}
	out, err := renderCSV(t, data, nil, []string{"ID", "Score"})
	require.NoError(t, err)

	lines := strings.Split(strings.TrimRight(out, "\n"), "\n")
	require.Len(t, lines, 3)
	assert.Equal(t, "ID,Score", lines[0])
	assert.Equal(t, "1,10", lines[1])
	assert.Equal(t, "2,20", lines[2])
}

func TestCSV_UnknownColumn(t *testing.T) {
	data := []csvRow{{ID: "1", Name: "alpha", Score: 10}}
	_, err := renderCSV(t, data, nil, []string{"Bogus"})
	require.Error(t, err)
	assert.Contains(t, err.Error(), "Bogus")
}

func TestCSV_RegisteredOnDefault(t *testing.T) {
	f, ok := output.Default.Lookup(output.CSV)
	require.True(t, ok)
	assert.Equal(t, "csv", f.Key())

	exts := output.Default.ExtensionMap()
	assert.Equal(t, "csv", exts[".csv"], ".csv extension must map to csv formatter")
}

func TestCSV_SingleStruct(t *testing.T) {
	out, err := renderCSV(t, csvRow{ID: "1", Name: "alpha", Score: 10}, nil, nil)
	require.NoError(t, err)

	lines := strings.Split(strings.TrimRight(out, "\n"), "\n")
	require.Len(t, lines, 2, "header + 1 row")
	assert.Equal(t, "ID,Name,Score", lines[0])
	assert.Equal(t, "1,alpha,10", lines[1])
}

func TestCSV_DelimiterMustBeSingleChar(t *testing.T) {
	data := []csvRow{{ID: "1", Name: "alpha", Score: 10}}
	_, err := renderCSV(t, data, []string{"delimiter=;;"}, nil)
	require.Error(t, err)
}

// --- CR/LF preservation -------------------------------------------------

// adversarialCSVRow pins the quoting rule byte-for-byte. Every field is a
// separate hazard: the delimiter, an internal quote, an embedded LF, a
// leading space, a trailing space, empty, a tab, and a LONE CR.
type adversarialCSVRow struct {
	A string `table:"a"`
	B string `table:"b"`
	C string `table:"c"`
	D string `table:"d"`
	E string `table:"e"`
	F string `table:"f"`
	G string `table:"g"`
	H string `table:"h"`
	I string `table:"i"`
}

func adversarialCSVData() []adversarialCSVRow {
	return []adversarialCSVRow{{
		A: "plain",
		B: "with,comma",
		C: `with"quote`,
		D: "with\nnewline",
		E: " leading space",
		F: "trailing ",
		G: "",
		H: "with\ttab",
		I: "with\rcr",
	}}
}

func adversarialCSVValues() []string {
	return []string{
		"plain", "with,comma", `with"quote`, "with\nnewline",
		" leading space", "trailing ", "", "with\ttab", "with\rcr",
	}
}

// A field holding CR and/or LF is quoted and its bytes survive verbatim, in
// BOTH line-ending modes and BOTH quoting paths. RFC 4180 lists CR and LF as
// separate alternatives inside `escaped`, so a bare CR between quotes is
// legal; dropping it is silent data loss, and promoting LF to CRLF rewrites
// the caller's value. The record terminator is the only place CRLF belongs.
func TestCSV_PreservesCRAndLFVerbatim(t *testing.T) {
	for _, tc := range []struct {
		name string
		opts []string
		eol  string
	}{
		{"lf", []string{"no-header=true"}, "\n"},
		{"crlf", []string{"no-header=true", "crlf=true"}, "\r\n"},
		{"lf/quote-all", []string{"no-header=true", "quote-all=true"}, "\n"},
		{"crlf/quote-all", []string{"no-header=true", "quote-all=true", "crlf=true"}, "\r\n"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			out, err := renderCSV(t, adversarialCSVData(), tc.opts, nil)
			require.NoError(t, err)

			assert.Contains(t, out, "\"with\rcr\"", "lone CR must be quoted and preserved, never dropped")
			assert.Contains(t, out, "\"with\nnewline\"", "embedded LF must stay LF inside the field")
			assert.NotContains(t, out, "withcr", "CR must not be silently discarded")
			assert.True(t, strings.HasSuffix(out, tc.eol), "record terminator")
		})
	}
}

// The real acceptance criterion: encode -> decode -> identical. Byte-equality
// alone would be satisfied by every runtime agreeing on lossy output.
func TestCSV_RoundTripsAdversarialRow(t *testing.T) {
	for _, tc := range []struct {
		name string
		opts []string
	}{
		{"lf", []string{"no-header=true"}},
		{"crlf", []string{"no-header=true", "crlf=true"}},
		{"lf/quote-all", []string{"no-header=true", "quote-all=true"}},
		{"crlf/quote-all", []string{"no-header=true", "quote-all=true", "crlf=true"}},
		{"semicolon", []string{"no-header=true", "delimiter=;"}},
	} {
		t.Run(tc.name, func(t *testing.T) {
			out, err := renderCSV(t, adversarialCSVData(), tc.opts, nil)
			require.NoError(t, err)

			r := csv.NewReader(strings.NewReader(out))
			r.Comma = ','
			for _, o := range tc.opts {
				if o == "delimiter=;" {
					r.Comma = ';'
				}
			}
			recs, err := r.ReadAll()
			require.NoError(t, err, "output must be parseable by encoding/csv")
			require.Len(t, recs, 1, "one input row must decode to exactly one record")
			assert.Equal(t, adversarialCSVValues(), recs[0])
		})
	}
}

// Go's own encoding/csv reader normalises a CRLF *inside* a quoted field to a
// bare LF, so a writer that promotes LF->CRLF in-field looks lossless on a
// Go-only round-trip while still having rewritten the bytes. Assert on the
// bytes directly so the promotion cannot hide.
func TestCSV_CRLFModeDoesNotPromoteInFieldLF(t *testing.T) {
	type note struct {
		Note string `table:"Note"`
	}
	out, err := renderCSV(t, []note{{Note: "a\nb"}}, []string{"no-header=true", "crlf=true"}, nil)
	require.NoError(t, err)
	assert.Equal(t, "\"a\nb\"\r\n", out,
		"in-field LF stays LF; only the record terminator is CRLF")
}

// Go quotes on unicode.IsSpace of the first rune, not merely a literal space.
func TestCSV_QuotesAnyLeadingUnicodeSpace(t *testing.T) {
	type note struct {
		Note string `table:"Note"`
	}
	for _, tc := range []struct {
		name, in, want string
	}{
		{"tab", "\tlead", "\"\tlead\"\n"},
		{"space", " lead", "\" lead\"\n"},
		{"nbsp", "\u00a0lead", "\"\u00a0lead\"\n"},
		{"vtab", "\vlead", "\"\vlead\"\n"},
		{"trailing space stays bare", "trail ", "trail \n"},
		{"plain stays bare", "plain", "plain\n"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			out, err := renderCSV(t, []note{{Note: tc.in}}, []string{"no-header=true"}, nil)
			require.NoError(t, err)
			assert.Equal(t, tc.want, out)
		})
	}
}

// `\.` alone on a line terminates a PostgreSQL COPY stream; encoding/csv
// quotes it defensively and so must we.
func TestCSV_QuotesPostgresCopySentinel(t *testing.T) {
	type note struct {
		Note string `table:"Note"`
	}
	out, err := renderCSV(t, []note{{Note: `\.`}}, []string{"no-header=true"}, nil)
	require.NoError(t, err)
	assert.Equal(t, "\"\\.\"\n", out)
}

File: sdk/ts/test/output/csv.test.ts
import { describe, it, expect, vi } from 'vitest';
import { csvFormatter } from '../../src/output/formatters/csv';
import { defaultRegistry } from '../../src/output/registry';
import '../../src/output/builtins';

function makeStream() {
  const chunks: string[] = [];
  const stream = {
    write: vi.fn((c: string) => {
      chunks.push(c);
      return true;
    }),
  } as unknown as NodeJS.WritableStream;
  return { stream, captured: () => chunks.join('') };
}

const rows = [
  { id: '1', name: 'Alice', notes: 'a, b' },
  { id: '2', name: 'Bob', notes: '"quoted"' },
];

describe('csv — defaults', () => {
  it('emits header + rows with comma delimiter and CRLF default off', () => {
    const s = makeStream();
    csvFormatter.render(
      s.stream,
      rows,
      { delimiter: ',', 'no-header': false, 'quote-all': false, crlf: false },
      [],
    );
    const out = s.captured();
    expect(out.split('\n')[0]).toBe('id,name,notes');
    expect(out).toContain('1,Alice,"a, b"');
    expect(out).toContain('2,Bob,"""quoted"""');
    expect(out).not.toContain('\r\n');
  });
});

describe('csv — delimiter override', () => {
  it('uses semicolon when set', () => {
    const s = makeStream();
    csvFormatter.render(
      s.stream,
      rows,
      { delimiter: ';', 'no-header': false, 'quote-all': false, crlf: false },
      [],
    );
    expect(s.captured().split('\n')[0]).toBe('id;name;notes');
  });

  it('throws on multi-char delimiter', () => {
    const s = makeStream();
    expect(() =>
      csvFormatter.render(
        s.stream,
        rows,
        { delimiter: '||', 'no-header': false, 'quote-all': false, crlf: false },
        [],
      ),
    ).toThrow(/exactly one character/);
  });
});

describe('csv — no-header', () => {
  it('omits header row', () => {
    const s = makeStream();
    csvFormatter.render(
      s.stream,
      rows,
      { delimiter: ',', 'no-header': true, 'quote-all': false, crlf: false },
      [],
    );
    expect(s.captured().split('\n')[0]).toBe('1,Alice,"a, b"');
  });
});

describe('csv — quote-all', () => {
  it('quotes every field', () => {
    const s = makeStream();
    csvFormatter.render(
      s.stream,
      [{ a: '1', b: 'x' }],
      { delimiter: ',', 'no-header': false, 'quote-all': true, crlf: false },
      [],
    );
    const out = s.captured();
    expect(out).toContain('"a","b"');
    expect(out).toContain('"1","x"');
  });
});

describe('csv — crlf', () => {
  it('uses CRLF line endings', () => {
    const s = makeStream();
    csvFormatter.render(
      s.stream,
      [{ a: '1' }],
      { delimiter: ',', 'no-header': false, 'quote-all': false, crlf: true },
      [],
    );
    expect(s.captured()).toContain('\r\n');
  });
});

describe('csv — empty input', () => {
  it('produces no output', () => {
    const s = makeStream();
    csvFormatter.render(s.stream, [], {}, []);
    expect(s.captured()).toBe('');
  });
});

describe('csv — cols subset', () => {
  it('filters to requested columns', () => {
    const s = makeStream();
    csvFormatter.render(s.stream, rows, {}, ['id', 'name']);
    const out = s.captured();
    expect(out.split('\n')[0]).toBe('id,name');
    expect(out).not.toContain('notes');
  });
});

describe('csv — registry', () => {
  it('registered as "csv" with .csv extension on defaultRegistry', () => {
    const f = defaultRegistry.lookup('csv');
    expect(f).toBe(csvFormatter);
    expect(defaultRegistry.extensionMap().get('.csv')).toBe('csv');
  });
});

// --- CR/LF preservation -------------------------------------------------

// The adversarial row that pins csv quoting rules byte-for-byte. Every field
// is a separate hazard: the delimiter, an internal quote, an embedded LF, a
// leading space, a trailing space, empty, a tab, and a LONE CR.
const adversarialRow = {
  a: 'plain',
  b: 'with,comma',
  c: 'with"quote',
  d: 'with\nnewline',
  e: ' leading space',
  f: 'trailing ',
  g: '',
  h: 'with\ttab',
  i: 'with\rcr',
};

const adversarialValues = [
  'plain',
  'with,comma',
  'with"quote',
  'with\nnewline',
  ' leading space',
  'trailing ',
  '',
  'with\ttab',
  'with\rcr',
];

/**
 * A minimal RFC 4180 reader. The package ships no csv parser, so round-trip
 * is proved against a decoder written to the grammar directly:
 * `escaped = DQUOTE *(TEXTDATA / COMMA / CR / LF / 2DQUOTE) DQUOTE`.
 * Records split only on an UNQUOTED line ending.
 */
function decodeCsv(input: string, delim = ','): string[][] {
  const records: string[][] = [];
  let record: string[] = [];
  let field = '';
  let inQuotes = false;
  let pending = false;

  for (let i = 0; i < input.length; i++) {
    const c = input[i];
    pending = true;
    if (inQuotes) {
      if (c === '"') {
        if (input[i + 1] === '"') {
          field += '"';
          i++;
        } else {
          inQuotes = false;
        }
      } else {
        field += c;
      }
      continue;
    }
    if (c === '"' && field === '') {
      inQuotes = true;
    } else if (c === delim) {
      record.push(field);
      field = '';
    } else if (c === '\r' || c === '\n') {
      if (c === '\r' && input[i + 1] === '\n') i++;
      record.push(field);
      field = '';
      records.push(record);
      record = [];
      pending = false;
    } else {
      field += c;
    }
  }
  if (pending || field !== '') {
    record.push(field);
    records.push(record);
  }
  return records;
}

describe('csv — CR/LF preservation', () => {
  const modes: [string, boolean, boolean][] = [
    ['lf', false, false],
    ['crlf', true, false],
    ['lf/quote-all', false, true],
    ['crlf/quote-all', true, true],
  ];

  // The load-bearing case: in CRLF mode csv-stringify left the embedded LF
  // UNQUOTED, so one 9-field record decoded as two records. That is not a
  // parity question — it is structurally invalid CSV.
  it.each(modes)('quotes and preserves CR/LF verbatim (%s)', (_label, crlf, quoteAll) => {
    const s = makeStream();
    csvFormatter.render(
      s.stream,
      [adversarialRow],
      { delimiter: ',', 'no-header': true, 'quote-all': quoteAll, crlf },
      [],
    );
    const out = s.captured();
    expect(out).toContain('"with\rcr"');
    expect(out).toContain('"with\nnewline"');
    expect(out).toContain('" leading space"');
    expect(out.endsWith(crlf ? '\r\n' : '\n')).toBe(true);
  });

  // Round-trip is the acceptance criterion, not byte-equality: byte-equality
  // alone would be satisfied by every runtime agreeing on lossy output.
  it.each(modes)('round-trips the adversarial row (%s)', (_label, crlf, quoteAll) => {
    const s = makeStream();
    csvFormatter.render(
      s.stream,
      [adversarialRow],
      { delimiter: ',', 'no-header': true, 'quote-all': quoteAll, crlf },
      [],
    );
    const recs = decodeCsv(s.captured());
    expect(recs).toHaveLength(1);
    expect(recs[0]).toEqual(adversarialValues);
  });

  it('changes only the record terminator in crlf mode', () => {
    const s = makeStream();
    csvFormatter.render(
      s.stream,
      [{ v: 'a\nb' }],
      { delimiter: ',', 'no-header': true, 'quote-all': false, crlf: true },
      [],
    );
    expect(s.captured()).toBe('"a\nb"\r\n');
  });

  it.each([
    ['tab', '\tlead', '"\tlead"\n'],
    ['space', ' lead', '" lead"\n'],
    ['nbsp', '\u00a0lead', '"\u00a0lead"\n'],
    ['vtab', '\vlead', '"\vlead"\n'],
    ['trailing space stays bare', 'trail ', 'trail \n'],
    ['plain stays bare', 'plain', 'plain\n'],
  ])('quotes any leading unicode space (%s)', (_label, input, want) => {
    const s = makeStream();
    csvFormatter.render(
      s.stream,
      [{ v: input }],
      { delimiter: ',', 'no-header': true, 'quote-all': false, crlf: false },
      [],
    );
    expect(s.captured()).toBe(want);
  });

  // `\.` alone on a line terminates a PostgreSQL COPY stream.
  it('quotes the postgres COPY sentinel', () => {
    const s = makeStream();
    csvFormatter.render(
      s.stream,
      [{ v: '\\.' }],
      { delimiter: ',', 'no-header': true, 'quote-all': false, crlf: false },
      [],
    );
    expect(s.captured()).toBe('"\\."\n');
  });
});
